src/Features/Core/Portable/AddDebuggerDisplay/AbstractAddDebuggerDisplayCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,117 @@
+
+        private async Task<Document> ApplyAsync(Document document, TTypeDeclarationSyntax type, CancellationToken cancellationToken)
+        {
+            var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var newAttribute = generator
+                .Attribute("System.Diagnostics.DebuggerDisplayAttribute", generator.LiteralExpression("{ToString(),nq}"))
+                .WithAdditionalAnnotations(Simplifier.Annotation);
+
+            syntaxRoot = syntaxRoot.ReplaceNode(type, generator.AddAttributes(type, newAttribute).WithAdditionalAnnotations(s_trackingAnnotation));
+
+            syntaxRoot = await EnsureNamespaceImportAsync(
+                document,
+                generator,
+                syntaxRoot,
+                contextLocation: syntaxRoot.GetAnnotatedNodes(s_trackingAnnotation).Single(),
+                "System.Diagnostics",
+                cancellationToken).ConfigureAwait(false);
+
+            return document.WithSyntaxRoot(syntaxRoot);
+        }
+
+        private static async Task<SyntaxNode> EnsureNamespaceImportAsync(
+            Document document,
+            SyntaxGenerator generator,
+            SyntaxNode syntaxRoot,
+            SyntaxNode contextLocation,
+            string namespaceName,
+            CancellationToken cancellationToken)
+        {
+            var compilation = await document.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+            var importsService = document.Project.LanguageServices.GetRequiredService<IAddImportsService>();
+            var newImport = generator.NamespaceImportDeclaration(namespaceName);
+
+            if (importsService.HasExistingImport(compilation, syntaxRoot, contextLocation, newImport))
+            {
+                return syntaxRoot;
+            }
+
+            var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var placeSystemNamespaceFirst = optionSet.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, document.Project.Language);
+
+            return importsService.AddImport(
+                compilation,
+                syntaxRoot,
+                contextLocation,
+                newImport,
+                placeSystemNamespaceFirst);
+        }
+
+        protected abstract bool HasDebuggerDisplayAttribute(TTypeDeclarationSyntax typeDeclaration);
+
+        protected abstract bool IsToStringOverride(TMethodDeclarationSyntax methodDeclaration);
+
+        protected abstract bool DeclaresToStringOverride(TTypeDeclarationSyntax typeDeclaration);
+
+        protected bool IsDebuggerDisplayAttributeIdentifier(SyntaxToken name)
+        {
+            return name.ValueText == "DebuggerDisplay" || name.ValueText == "DebuggerDisplayAttribute";
+        }
+
+        private sealed class CodeAction : CodeActions.CodeAction.DocumentChangeAction
```

Comment:
can you call it MyCodeAction.  That way this doesn't come up first if i nav-to 'CodeAction'.  Thanks!

Reply:
Gross, but sure.